Sources/AsyncHTTPClient/ConnectionPool.swift

```diff
@@ -0,0 +1,600 @@
+                channel.pipeline.addSSLHandlerIfNeeded(for: self.key, tlsConfiguration: self.configuration.tlsConfiguration, handshakePromise: handshakePromise).flatMap {
+                    channel.pipeline.addHTTPClientHandlers(leftOverBytesStrategy: .forwardBytes)
+                }.map {
+                    let connection = Connection(key: self.key, channel: channel, parentPool: self.parentPool)
+                    connection.isLeased = true
+                    return connection
+                }
+            }.and(handshakePromise.futureResult).map { (arg) -> ConnectionPool.Connection in
+                let (connection, _) = arg
+                self.configureCloseCallback(of: connection)
+                return connection
+            }
+
+            connection.whenFailure { _ in
+                let action = self.stateLock.withLock {
+                    self.state.failedConnectionAction()
+                }
+                switch action {
+                case .makeConnectionAndComplete(let el, let promise):
+                    self.makeConnection(on: el).cascade(to: promise)
+                case .none:
+                    break
+                }
+            }
+
+            return connection
+        }
+
+        /// Adds a callback on connection close that asks the `state` what to do about this
+        ///
+        /// The callback informs the state about the event, and the state returns a
+        /// `ClosedConnectionRemoveAction` which instructs it about what it should do.
+        private func configureCloseCallback(of connection: Connection) {
+            connection.channel.closeFuture.whenComplete { result in
+                switch result {
+                case .success:
+                    let action = self.parentPool.connectionProvidersLock.withLock {
+                        self.stateLock.withLock { self.state.removeClosedConnection(connection) }
+                    }
+                    switch action {
+                    case .none:
+                        break
+                    case .makeConnectionAndComplete(let el, let promise):
+                        self.makeConnection(on: el).cascade(to: promise)
+                    }
+
+                case .failure(let error):
+                    preconditionFailure("Connection close future failed with error: \(error)")
+                }
+            }
+        }
+
+        func syncClose(requiresCleanClose: Bool) throws {
+            var closeError: Error?
+            let availableConnections = self.stateLock.withLock { () -> CircularBuffer<ConnectionPool.Connection> in
+                assert(!self.state.isClosed, "Calling syncClose on an already closed provider")
+                let waitersCopy = self.state.waiters
+                self.state.waiters.removeAll()
+                for waiter in waitersCopy {
+                    waiter.promise.fail(HTTPClientError.cancelled)
```

Comment:
Never called by unit tests, must check if that's normal or if additional tests should be added for this